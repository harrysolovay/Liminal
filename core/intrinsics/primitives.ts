import { assert } from "@std/assert"
import type { Type } from "../Type.ts"
import { declare } from "./declare.ts"

export { null_ as null }
const null_: Type<null> = declare({
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean> = declare({
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number> = declare({
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number> = declare({
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string> = declare({
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
