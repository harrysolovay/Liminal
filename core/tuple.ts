import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import type { Expand } from "../util/type_util.ts"

export function tuple<E extends Array<Type>>(
  ...elements: E
): Type<Expand<NativeTuple<E>>, {}, E[number]["P"]> {
  const { length } = elements
  const required = Array.from({ length }, (_0, i) => i.toString())
  return declare<NativeTuple<E>>()({
    name: "tuple",
    source: {
      factory: tuple,
      args: { elements },
    },
    subschema: (ref) => ({
      type: "object",
      properties: Object.fromEntries(
        Array.from({ length }, (_0, i) => [i, ref(elements[i]!)]),
      ),
      required,
      additionalProperties: false,
    }),
    visitor: (value, visit) =>
      Array.from(
        { length },
        (_0, i) => visit(value[i]!, elements[i]!, i),
      ) as never,
    assertRefinementsValid: () => {},
    assertRefinements: {},
  })
}

export type NativeTuple<E extends Array<Type>> = { [K in keyof E]: E[K]["T"] }
