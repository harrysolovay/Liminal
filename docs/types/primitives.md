# Primitive Types

```ts twoslash
import { L } from "liminal"
// ---cut---
L.boolean
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.number
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.integer
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.string
// ^?
```

<br />
