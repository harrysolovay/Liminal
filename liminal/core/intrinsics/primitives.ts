import { assert } from "@std/assert"
import { declare } from "../declare.ts"
import type { Type } from "../Type.ts"

export { null_ as null }
const null_: Type<null, never> = declare({
  jsonType: "null",
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean, never> = declare({
  jsonType: "boolean",
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number, never> = declare({
  jsonType: "integer",
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number, never> = declare({
  jsonType: "number",
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string, never> = declare({
  jsonType: "string",
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
