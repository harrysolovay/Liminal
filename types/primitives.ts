import { Ty } from "./Ty.ts"

export const none: Ty<null, never> = Ty((description) => ({
  type: "null",
  description,
}))

export const boolean: Ty<boolean, never> = Ty((description) => ({
  type: "boolean",
  description,
}))

export const number: Ty<number, never> = Ty((description) => ({
  type: "number",
  description,
}))

export const string: Ty<string, never> = Ty((description) => ({
  type: "string",
  description,
}))
