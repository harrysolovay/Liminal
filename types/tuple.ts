import { Ty } from "./Ty.ts"

export function tuple<F extends Ty[]>(
  ...elements: F
): Ty<{ [K in keyof F]: F[K]["T"] }, F[number]["P"], true> {
  const { length } = elements
  return Ty(
    (ref) => ({
      type: "object",
      properties: Object.fromEntries(Array.from({ length }, (_0, i) => [i, ref(elements[i]!)])),
      additionalProperties: false,
      required: Object.keys(Array.from({ length }, (_0, i) => i)),
    }),
    true,
    (v) => Array.from({ length }, (_0, i) => elements[i]![""].transform(v[i])) as never,
  )
}
