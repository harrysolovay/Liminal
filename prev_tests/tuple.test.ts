import { assertTySnapshot } from "test_util"
import { boolean, number, string } from "./primitives.ts"
import { tuple } from "./tuple.ts"

Deno.test("tuple", async (t) => {
  await assertTySnapshot(t, tuple(boolean, number, string))
})
