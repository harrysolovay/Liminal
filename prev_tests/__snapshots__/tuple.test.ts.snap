export const snapshot = {};

snapshot[`tuple 1`] = `
{
  additionalProperties: false,
  properties: {
    "0": {
      type: "boolean",
    },
    "1": {
      type: "number",
    },
    "2": {
      type: "string",
    },
  },
  required: [
    "0",
    "1",
    "2",
  ],
  type: "object",
}
`;

snapshot[`tuple 2`] = `
{
  additionalProperties: false,
  description: "One.",
  properties: {
    "0": {
      type: "boolean",
    },
    "1": {
      type: "number",
    },
    "2": {
      type: "string",
    },
  },
  required: [
    "0",
    "1",
    "2",
  ],
  type: "object",
}
`;

snapshot[`tuple 3`] = `
{
  additionalProperties: false,
  description: "Two. One.",
  properties: {
    "0": {
      type: "boolean",
    },
    "1": {
      type: "number",
    },
    "2": {
      type: "string",
    },
  },
  required: [
    "0",
    "1",
    "2",
  ],
  type: "object",
}
`;
